Add manual test option to package docs

barrettk opened the pull request:
This feature was previously allowed, and has been added back in

closes https://github.com/metrumresearchgroup/mrgvalidate/issues/99

barrettk committed:
Add manual test option to package docs
 - added recursive test

kyleam reviewed (changes requested):
If I run the first test in test-create-package-docs.R, which does _not_ pass a `man_test_dir`, I still see the manual test section in the plan:

<img width="838" alt="image" src="https://user-images.githubusercontent.com/1297788/217073801-5fa882e9-4d93-479e-9a36-4c7b973fdf0f.png">

On the other hand, I don't see that in the test results doc (which is good).  I suspect (but haven't verified) that that difference comes down to `make_testing_results()` having an `else` arm that sets `man_tests` to NULL:

https://github.com/metrumresearchgroup/mrgvalidate/blob/82ee233d57e852c9538ac38b1d004d873460af1c/R/make-testing-results.R#L60-L70

https://github.com/metrumresearchgroup/mrgvalidate/blob/82ee233d57e852c9538ac38b1d004d873460af1c/R/make-testing-plan.R#L61-L72

It'd probably also make sense to have `format_man_test_plan()` and `format_man_test_results` ignore the empty string case.  Right now they only ignore NULL and `length(x) < 1`.

https://github.com/metrumresearchgroup/mrgvalidate/blob/82ee233d57e852c9538ac38b1d004d873460af1c/R/make-testing-results.R#L101-L104

https://github.com/metrumresearchgroup/mrgvalidate/blob/82ee233d57e852c9538ac38b1d004d873460af1c/R/make-testing-plan.R#L119-L122

kyleam commented on tests/testthat/test-create-package-docs.R:
```diff
@@ -365,3 +365,83 @@ test_that("create_package_docs() changes test plan automated test section based
+
+
+
+test_that("create_package_docs() works with manual tests", {
```
As a side remark that I think is fine to not address in this series: even before this PR, there's a fair amount of duplicated test setup.  At some point, it might be worth trying to extract out some common bits so that it's easier to  see the main point of a given test.

For example, here is how this test compares to the first test in this file:

```diff
@@ -1,11 +1,11 @@
-test_that("create_package_docs() renders markdown", {
+test_that("create_package_docs() works with manual tests", {
   # set up clean docs output dir
   output_dir <- file.path(tempdir(), "mrgvalidate-create-validation-docs")
   if (fs::dir_exists(output_dir)) fs::dir_delete(output_dir)
   fs::dir_create(output_dir)
   on.exit({ fs::dir_delete(output_dir) })
 
-  specs <- readRDS(file.path(TEST_INPUTS_DIR, "specs.RDS")) %>% filter(!grepl("MAN", TestIds))
+  specs <- readRDS(file.path(TEST_INPUTS_DIR, "specs.RDS"))
   specs$StoryDescription[1] <- "story desc line 2\nLINE2!!"
 
   # Test that repeating TestIDs in traceability matrix are removed
@@ -21,6 +21,7 @@ test_that("create_package_docs() renders markdown", {
     specs = specs,
     release_notes_file = file.path(TEST_INPUTS_DIR, "release_notes_sample.md"),
     auto_test_dir = file.path(TEST_INPUTS_DIR, "validation-results-sample"),
+    man_test_dir = file.path(TEST_INPUTS_DIR, "manual-tests-sample"),
     output_dir = output_dir,
     write = TRUE,
     cleanup_rmd = FALSE
@@ -31,7 +32,8 @@ test_that("create_package_docs() renders markdown", {
 
   # get TestId's we expect to see
   test_ids <- c(
-    read_csv_test_results(file.path(TEST_INPUTS_DIR, "validation-results-sample"))$results %>% pull(TestId)
+    read_csv_test_results(file.path(TEST_INPUTS_DIR, "validation-results-sample"))$results %>% pull(TestId),
+    read_manual_test_results(file.path(TEST_INPUTS_DIR, "manual-tests-sample")) %>% pull(TestId)
   ) %>% unique()
 
 
```

barrettk replied:
Yeah I thought about that. I was using this for my own sake as a sort of integration test to make sure everything else still behaved normally. But yeah we definitely dont need most of that. Ill redo this test

barrettk commented:
> If I run the first test in test-create-package-docs.R, which does _not_ pass a `man_test_dir`, I still see the manual test section in the plan:

Ahh good point. I modified the templates to have them, but I need to change the helper functions for some of the templates to contain the section header (some already have this)

barrettk committed:
Fixed manual test bug (used to show in test plan package docs)

 - made recursive test less repetitive and more comprehensive

barrettk commented:
PR is ready for review again (took me a minute to get back to this). 

> It'd probably also make sense to have format_man_test_plan() and format_man_test_results ignore the empty string case. Right now they only ignore NULL and length(x) < 1.

I found that an empty string wasn't possible here (or at least something where `rlang::is_empty()` would ever be `TRUE`). The returned result should always be `NULL` if there are manual tests:

This will force it to be a tibble
```r
man_tests <- filter(tests, .data$test_type == "manual")
```

This will force it to either be a character string or `NULL`. Regardless, you shouldnt be able to get an empty character string (unless the manual test directory was empty or something? - haven't tested cases like this, but not sure that's high priority at the moment):
```r
  if (nrow(man_tests) != 0) {
    man_tests <- pull(man_tests, .data$man_test_content) %>%
      # Remove Results from manual tests (everything after run details)
      # might want a tryCatch here? (unsure if they're always in that format)
      # revisit this for updated manual sections
      map(~ {
        .x <- gsub("### Run Details.*", "",.x)
        str_trim(.x)
      }) %>%
      map(~ glue("\n{.x}\n\n"))
  }else{
    # Change to NULL if no manual tests exist (used in helper functions)
    man_tests <- NULL
  }
```

kyleam commented:
> > It'd probably also make sense to have format_man_test_plan() and format_man_test_results ignore the empty string case. Right now they only ignore NULL and length(x) < 1.
> 
> I found that an empty string wasn't possible here

Hmm the templates set the `man_tests` default to `""`.  For example

https://github.com/metrumresearchgroup/mrgvalidate/blob/28b390c63c8bc5235d75e53ab8bcb91bfe57e683/inst/templates/package/testing_results_package.Rmd#L8-L12

... and then call these functions directly:

https://github.com/metrumresearchgroup/mrgvalidate/blob/28b390c63c8bc5235d75e53ab8bcb91bfe57e683/inst/templates/package/testing_results_package.Rmd#L28

But based on you talking about the `make_testing_plan` logic, I guess you just mean it's not currently reachable given how the current code base calls `rmarkdown::render` with these template.  While that may be true (I haven't checked all the templates and render calls), it seems odd to me for these functions to be used by the templates but not handle the default values, relying on the calling site to follow an undocumented expectation.  And guarding against that default seems like a good (and easy) way to avoid an unintentional regression when tweaking the render calls later, especially since that regression might not be caught by the automatic tests (just like the issue I pointed out wasn't).

Not a blocker for this PR, more of a side remark.  Up to you if you to address it while in the area.

kyleam reviewed (approved):
Update looks good to me, and I confirmed manually that testing-plan-package_TEST.docx from the first test no longer has the empty manual tests section (though it'd be nice to include a regression test for this).

barrettk commented:
Ahh that's a fair point about the templates. As you mention, those never get evaluated, but I could change them all to `NULL` for consistency's sake. [`create_validation_templates()`](https://github.com/metrumresearchgroup/mrgvalidate/blob/main/R/create-templates.R) uses its own values. Ill update the default values for now.

barrettk committed:
change template defaults for auto and manual tests:
 - changed in the testing plan and results templates for both packages and metworx. No impact on tests (always overwritten), but changed for consistency
